velox/exec/VectorHasher.cpp

```diff
@@ -247,36 +246,43 @@ bool VectorHasher::makeValueIdsDecoded(
   auto values = decoded_.data<T>();
 
   bool success = true;
-  rows.applyToSelected([&](vector_size_t row) INLINE_LAMBDA {
+  int decodedBaseVisited = 0;
```

Comment:
This name confuses me. It seems to tracking number of hashes computed and cached. Would numCachedHashes be clearer?

Also, this counter is needed to exit the loop earlier, but I wonder how important this is. As long as we do not call analyzeValue more than once per value, the perf should be fine, no? Do you see any perf difference if decodedBaseVisited is removed? If not, perhaps, worth removing this variable to reduce complexity.

Reply:
Hi Masha, this variable tracks the number of base vector values visited.

A visit is either a hash in "success" mode, or an analyzeValue in "failure" mode. If in "failure" mode, there is no "compute and cache" action, thus, renaming to numCachedHashes would be misleading.

As for performance, yes, this variable is necessary because it is the basis of this optimization, that, if we already know the current "input batch" is failure, we do not want to spend any more time analyzing the remaining values than we have to. i.e, only analyze until all base values are visited.

Reply:
My impression is that the logic below ensures that analyzeValue is called only once per base row. Is this not sufficient? What is the performance difference when decodedBaseVisited is removed?

```
   } else {
      if (id == 0) {
        id = kUnmappable;
        decodedBaseVisited++;
        analyzeValue(values[baseIndex]);
      }
    }
```

Reply:
Let me adopt the rename suggestion.

Yes, this variable is needed. For example, if the batch is quite large but the base vector has only 2 values, without this variable, a lot of time will be wasted on simply iterating through the selectivity.

In fact the included benchmark demonstrates this. The batch has 30mil values, but base vector only has 300 values. In this case, without this variable, we will spend orders of magnitude more time. The reason is because iterating through the bit set itself is "not cheap" cpu instruction-wise, even if you don't analyze the values already analyzed.

Reply:
Got it. Appreciate the explanation. It would be great to have 2 benchmarks, one there we are able to exit early because we visited all base rows and one there we cannot do that. There should be a significant performance difference between the 2 benchmarks and that would make clear the point above. 

How common is it to have 30M row batches? I haven't seen these in practice. Usually, I see something like ~10K row batches.